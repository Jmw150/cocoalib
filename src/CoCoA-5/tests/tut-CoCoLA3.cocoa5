------------------------------------------------------------
-- TUTORIAL 3
------------------------------------------------------------
-- ZERO DIMENSIONAL AFFINE ALGEBRAS
------------------------------------------------------------

------------------------------------------------------------
-- MULTIPLICATION ENDOMORPHISMS
------------------------------------------------------------

use P ::= QQ[x,y,z];  -- "::=" for the special meaning of [..]
I := ideal(x*y-z, x^2-3, y^2-1);
R := P/I;
IsZeroDim(I); --> true
QB := QuotientBasis(I);   QB; --> [1, z, y, x] in P

use R; -- interpret the symbols x,y,z in R, the quotient ring
f := x +y -z^2; -- in R
-- multiplication by f:
1 * f; -- "(..)" reminds this is the class of ".."
z * f;
y * f;
x * f;

-- it is easier to think in terms of polynomials:
use P; -- interpret the symbols x,y,z in P, the base ring
f := x +y -z^2; -- in P
NF(1 * f, I);
NF(z * f, I);
NF(y * f, I);
NF(x * f, I);

-- get the coefficients of the linear combinations
QB;
coefficients(NF(x * f, I), QB);
-- putting everything together:
-- matrix of the MULTIPLICATION ENDOMORPHISM by f on R
MultiplicationMat(f, I);

-- MULTIPLICATION FAMILY of R: generated by
S := [ MultiplicationMat(X,I) | X in indets(P) ];
indent(S);
-- indeed S are commuting endomorphisms:
for i := 1 to NumIndets(P) do
  for j := i+1 to NumIndets(P) do
    print i, j, ": ", S[i]*S[j] = S[j]*S[i], "  ";
  endfor;
endfor;

------------------------------------------------------------
-- CHINESE REMAINDER THEOREM
------------------------------------------------------------
-- Q1, Q2, Q3 comaximal ideals:
--   P/(Q1 ∩ Q2 ∩ Q3) ≅ P/Q1 x P/Q2 x P/Q3

Q1 := ideal(x-z+3, y-5*z+1, (z-1)^2);
Q2 := ideal(z^2+3*z, y-2*z-1, z^2+11);
Q3 := ideal(x^2-2, y-1, z^2+10);

-- are they comaximal?
1 IsIn Q1+Q2;
1 IsIn Q1+Q3;
1 IsIn Q2+Q3;

------------------
Q := [Q1, Q2, Q3];
E := NewList(3);
------------------------------
J := product(WithoutNth(Q,1));
repr1 := GenRepr(one(P), Q[1] + J); -- 1 = q+j
-- NOTE: GenRepr is a costly function

E[1] := ScalarProduct(gens(J), last(repr1, NumGens(J))); -- j = 1-q
E[1];
E[1] - 1 IsIn Q[1];  -- E1 = 1 mod Q[1]
E[1] IsIn Q[2];      -- E1 = 0 mod Q[2]
E[1] IsIn Q[3];      -- E1 = 0 mod Q[3]

------------------------------
J := product(WithoutNth(Q,2));
repr1 := GenRepr(one(P), Q[2] + J);
E[2] := ScalarProduct(gens(J), last(repr1, NumGens(J)));
E[2];

------------------------------
J := product(WithoutNth(Q,3));
repr1 := GenRepr(one(P), Q[3] + J);
E[3] := ScalarProduct(gens(J), last(repr1, NumGens(J)));
E[3];

indent(E);
sum(E);
NF(sum(E), IntersectList(Q));

-- EXERCISE:  Try with other 0-dimensional ideals
-- EXERCISE:  Write function ChineseRemainderTut

------------------------------------------------------------
-- THE HEURISTIC ALGORITHM for the PRIMARY DECOMPOSITION
------------------------------------------------------------

use P;
J1 := ideal(x^2-2, (y-2)^3, z-3);
J2 := ideal(x-1, y^3-3, (z-5)^2);
I := intersect(J1, J2);
I;

L := 1234*x - 9876*y + 555*z;
mi_L := MinPoly(MultiplicationMat(L,I), z);   mi_L;
-- it may be computed with other strategies (ask Elisa Palezzato ;-)
MinPolyQuotDef(L, I, z);
MinPolyQuotElim(L, I, z);

Facs := factor(mi_L);
p := Facs.factors;  m := Facs.multiplicities;
p; m;

-- many ways to evaluate p in L:
-- (1) let CoCoA work it out for you:
subst(p[1], z, L);
-- (2) function "eval"  (remember that z is the 3rd indet)
eval(p[1],[0,0,L]);
-- (3) define the evaluation homomorphism (works across rings)
phi := PolyAlgebraHom(P, P, [0,0,L]); -- domain, codomain, images
phi(p[1]);

Q1 := I + ideal(phi(p[1])^m[1]);
IsPrimary(Q1); -- we will see how to do it in Tutorial4
--> true!

Q2 := I + ideal(phi(p[2])^m[2]);
IsPrimary(Q2); 
--> true!

-- then [Q1, Q2]  is the primary decomposition of I 
Q1; Q2;  ---> hmmmmm....
GBasis(Q1);
GBasis(Q2);

Q1 = J2;   Q2 = J1;

-- we could do:
Q := [ I + ideal(phi(p[i])^m[i]) | i in 1..len(p)];

-- EXERCISE:  Try with other 0-dimensional ideals
-- EXERCISE:  Compare timings of the single operations
-- EXERCISE:  Write function PartialDec for a partial decomposition
-- EXERCISE:  Write function PrimaryDecTut doing all the checks
-- EXERCISE:  Compare reliability and timings
----------------------------------------------------------------------
-- COMPUTING THE RADICAL

indent(I);
J := I;

EF := eigenfactors(MultiplicationMat(x,J), x);  -- same indet
EF;
g := product(EF);
g;
J := I + ideal(g);
deg(g);  multiplicity(P/J); --> different: another indet

EF := eigenfactors(MultiplicationMat(y,J), y);
EF;
g := product(EF);
g;
J := J + ideal(g);  -- now the ideal is "simpler"
deg(g);  multiplicity(P/J); --> different: another indet

EF := eigenfactors(MultiplicationMat(z,J), z);
EF;
g := product(EF);
g;
J := J + ideal(g);
deg(g);  multiplicity(P/J); --> different. But we are done!

-------------------------------
-- an experiment
indent(I);
J := I;

L := 123*x + 234*y + 456*z;
EF := eigenfactors(MultiplicationMat(L,J), x);
EF := subst(EF, x, L);  -- is in J
g := product(EF);
g;
J := I + ideal(g);
deg(g);  multiplicity(P/J); --> the same!
-- DONE!

-- EXERCISE: try with other 0-dimensional ideals.
-- EXERCISE: write a function RadicalTut which computes the radical.
----------------------------------------------------------------------
