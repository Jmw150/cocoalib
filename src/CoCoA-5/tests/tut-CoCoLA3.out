true
[1,  z,  y,  x]
(x +y -3)
(x +3*y -3*z)
(-3*y +z +1)
(-3*x +z +3)
x +y -3
x +3*y -3*z
-3*y +z +1
-3*x +z +3
[1,  z,  y,  x]
[3,  1,  0,  -3]
matrix(QQ,
 [[-3, 0, 1, 3],
  [0, -3, 1, 1],
  [1, 3, -3, 0],
  [1, 1, 0, -3]])
[
  matrix(QQ,
 [[0, 0, 0, 3],
  [0, 0, 1, 0],
  [0, 3, 0, 0],
  [1, 0, 0, 0]]),
  matrix(QQ,
 [[0, 0, 1, 0],
  [0, 0, 0, 1],
  [1, 0, 0, 0],
  [0, 1, 0, 0]]),
  matrix(QQ,
 [[0, 3, 0, 0],
  [1, 0, 0, 0],
  [0, 0, 0, 3],
  [0, 0, 1, 0]])
]
12: true  13: true  23: true  true
true
true
(-2/121)*z^3 +(13/121)*z^2 +(-20/121)*z +130/121
true
true
true
(109/1176)*z^4 +(-121/3528)*y*z^2 +(865/882)*z^2 +(-1331/3528)*y +2021/3528
(2/121)*z^3 +(-13/121)*z^2 +(20/121)*z -9/121
[
  (-2/121)*z^3 +(13/121)*z^2 +(-20/121)*z +130/121,
  (109/1176)*z^4 +(-121/3528)*y*z^2 +(865/882)*z^2 +(-1331/3528)*y +2021/3528,
  (2/121)*z^3 +(-13/121)*z^2 +(20/121)*z -9/121
]
(109/1176)*z^4 +(-121/3528)*y*z^2 +(865/882)*z^2 +(-1331/3528)*y +5549/3528
1
ideal(x*z -3*x -z +3,  x^2 +(-1/4)*z^2 +(5/2)*z -29/4,  z^3 -13*z^2 +55*z -75,  y^2*z^2 +(-2/3)*y^3 -10*y^2*z -2*y*z^2 +25*y^2 +20*y*z +(5/6)*z^2 -50*y +(-25/3)*z +137/6,  y^3*z -3*y^3 -3*z +9,  y^4 -4*y^3 +(19/4)*y*z^2 +(-95/2)*y*z -7*z^2 +(463/4)*y +70*z -163,  x*y^3 -6*x*y^2 -y^3 +12*x*y +6*y^2 -8*x -12*y +8)
z^12 +84468*z^11 +2528650026*z^10 +30516512329492*z^9 +359291615328480375*z^8 +16731316023841937907240*z^7 +387095858802788137374947564*z^6 +3534852277454951723748359926920*z^5 +26367601905008197629356178266103999*z^4 +776576883337465756779672742594448411236*z^3 +13542823597271725413509179492878394368725514*z^2 +100266818901856595158865598758177724023748202980*z +271741657833857615206805916555341970467449413786793
z^12 +84468*z^11 +2528650026*z^10 +30516512329492*z^9 +359291615328480375*z^8 +16731316023841937907240*z^7 +387095858802788137374947564*z^6 +3534852277454951723748359926920*z^5 +26367601905008197629356178266103999*z^4 +776576883337465756779672742594448411236*z^3 +13542823597271725413509179492878394368725514*z^2 +100266818901856595158865598758177724023748202980*z +271741657833857615206805916555341970467449413786793
z^12 +84468*z^11 +2528650026*z^10 +30516512329492*z^9 +359291615328480375*z^8 +16731316023841937907240*z^7 +387095858802788137374947564*z^6 +3534852277454951723748359926920*z^5 +26367601905008197629356178266103999*z^4 +776576883337465756779672742594448411236*z^3 +13542823597271725413509179492878394368725514*z^2 +100266818901856595158865598758177724023748202980*z +271741657833857615206805916555341970467449413786793
[z^3 -12027*z^2 +48216243*z +2825345147399,  z^2 +36174*z +324094057]
[2,  3]
1879080904*x^3 -45116214768*x^2*y +361075961952*x*y^2 -963259373376*y^3 +2535388740*x^2*z -40582656720*x*y*z +162396401040*y^2*z +1140308550*x*z^2 -9126164700*y*z^2 +170953875*z^3 -18314186412*x^2 +293145713136*x*y -1173057967152*y^2 -16473862980*x*z +131844303720*y*z -3704616675*z^2 +59498843862*x -476183615868*y +26760014865*z +2825345147399
1879080904*x^3 -45116214768*x^2*y +361075961952*x*y^2 -963259373376*y^3 +2535388740*x^2*z -40582656720*x*y*z +162396401040*y^2*z +1140308550*x*z^2 -9126164700*y*z^2 +170953875*z^3 -18314186412*x^2 +293145713136*x*y -1173057967152*y^2 -16473862980*x*z +131844303720*y*z -3704616675*z^2 +59498843862*x -476183615868*y +26760014865*z +2825345147399
1879080904*x^3 -45116214768*x^2*y +361075961952*x*y^2 -963259373376*y^3 +2535388740*x^2*z -40582656720*x*y*z +162396401040*y^2*z +1140308550*x*z^2 -9126164700*y*z^2 +170953875*z^3 -18314186412*x^2 +293145713136*x*y -1173057967152*y^2 -16473862980*x*z +131844303720*y*z -3704616675*z^2 +59498843862*x -476183615868*y +26760014865*z +2825345147399
true
true
ideal(x*z -3*x -z +3,  x^2 +(-1/4)*z^2 +(5/2)*z -29/4,  z^3 -13*z^2 +55*z -75,  y^2*z^2 +(-2/3)*y^3 -10*y^2*z -2*y*z^2 +25*y^2 +20*y*z +(5/6)*z^2 -50*y +(-25/3)*z +137/6,  y^3*z -3*y^3 -3*z +9,  y^4 -4*y^3 +(19/4)*y*z^2 +(-95/2)*y*z -7*z^2 +(463/4)*y +70*z -163,  x*y^3 -6*x*y^2 -y^3 +12*x*y +6*y^2 -8*x -12*y +8,  3530945043777457216*x^6 -169554035262623180544*x^5*y +3392454724987168823040*x^4*y^2 -36200845882196952238080*x^3*y^3 +217293083832603586083840*x^2*y^4 -695619609702039875579904*x*y^5 +927868620396724177637376*y^6 +9528401131101241920*x^5*z -381290476380696374400*x^4*y*z +6103119521451794803200*x^3*y^2*z -48844739379139323724800*x^2*y^3*z +195458122410202577433600*x*y^4*z -312859711008615989422080*y^5*z +10713660104864646000*x^4*z^2 -342976036290577776000*x^3*y*z^2 +4117380066133402896000*x^2*y^2*z^2 -21968257986565903296000*x*y^3*z^2 +43954318451240855136000*y^4*z^2 +6424723239546060000*x^3*z^3 -154255834798436520000*x^2*y*z^3 +1234546697300939280000*x*y^2*z^3 -3293458450173980640000*y^3*z^3 +2167172648671837500*x^2*z^4 -34688812120393950000*x*y*z^4 +138811470219210150000*y^2*z^4 +389880330636262500*x*z^5 -3120306438706425000*y*z^5 +29225227377515625*z^6 -68827675918170952896*x^5 +2754222558216597774720*x^4*y -44085416507207649308160*x^3*y^2 +352826234542287475338240*x^2*y^3 -1411876779716220059335680*x*y^4 +2259918164745119822690304*y^5 -154778606704152669600*x^4*z +4954922268428563257600*x^3*y*z -59483159225689413369600*x^2*y^2*z +317372058623937680409600*x*y^3*z -635001306922611128793600*y^4*z -139225489012649484000*x^3*z^2 +3342765630848281128000*x^2*y*z^2 -26752960591780894992000*x*y^2*z^2 +71370134739175612896000*y^3*z^2 -62617622692074930000*x^2*z^3 +1002287912004752040000*x*y*z^3 -4010776101685142280000*y^2*z^3 -14081353563250237500*x*z^4 +112696473088054575000*y*z^4 -1266637151961731250*z^5 +559015706555809056240*x^4 -17895750787504603693440*x^3*y +214836022825035007386240*x^2*y^2 -1146256381102131676362240*x*y^3 +2293441657164637851651840*y^4 +1005684658471552759200*x^3*z -24146211556884250526400*x^2*y*z +193247962184593888329600*x*y^2*z -515536703007846904684800*y^3*z +678470403709536201000*x^2*z^2 -10859924970883921428000*x*y*z^2 +43457301058529014596000*y^2*z^2 +203431158324577305000*x*z^3 -1628108686882921770000*y*z^3 +22873641181146759375*z^4 +8438758391798752331104*x^3 -202612263883479282063168*x^2*y +1621554876915106474599552*x*y^2 -4325898423666556332562176*y^3 +11386169142905123688240*x^2*z -182252522618040521142720*x*y*z +729305475435886623503040*y^2*z +5121008001873860329800*x*z^2 -40984663716779776837200*y*z^2 +767736207736356694500*z^3 -99947682994495150809732*x^2 +1599810886958888345862096*x*y -6401836434200154499081872*y^2 -89904317766523190760780*x*z +719525966176809588292920*y*z -20217543095794315588425*z^2 +336209539562704952829876*x -2690766136727126510654664*y +151212556286305712172270*z +7982575201931077036465201)
ideal(x*z -3*x -z +3,  x^2 +(-1/4)*z^2 +(5/2)*z -29/4,  z^3 -13*z^2 +55*z -75,  y^2*z^2 +(-2/3)*y^3 -10*y^2*z -2*y*z^2 +25*y^2 +20*y*z +(5/6)*z^2 -50*y +(-25/3)*z +137/6,  y^3*z -3*y^3 -3*z +9,  y^4 -4*y^3 +(19/4)*y*z^2 +(-95/2)*y*z -7*z^2 +(463/4)*y +70*z -163,  x*y^3 -6*x*y^2 -y^3 +12*x*y +6*y^2 -8*x -12*y +8,  3530945043777457216*x^6 -169554035262623180544*x^5*y +3392454724987168823040*x^4*y^2 -36200845882196952238080*x^3*y^3 +217293083832603586083840*x^2*y^4 -695619609702039875579904*x*y^5 +927868620396724177637376*y^6 +9528401131101241920*x^5*z -381290476380696374400*x^4*y*z +6103119521451794803200*x^3*y^2*z -48844739379139323724800*x^2*y^3*z +195458122410202577433600*x*y^4*z -312859711008615989422080*y^5*z +10713660104864646000*x^4*z^2 -342976036290577776000*x^3*y*z^2 +4117380066133402896000*x^2*y^2*z^2 -21968257986565903296000*x*y^3*z^2 +43954318451240855136000*y^4*z^2 +6424723239546060000*x^3*z^3 -154255834798436520000*x^2*y*z^3 +1234546697300939280000*x*y^2*z^3 -3293458450173980640000*y^3*z^3 +2167172648671837500*x^2*z^4 -34688812120393950000*x*y*z^4 +138811470219210150000*y^2*z^4 +389880330636262500*x*z^5 -3120306438706425000*y*z^5 +29225227377515625*z^6 +310522867131942635328*x^5 -12425947470806586168960*x^4*y +198895716728826329018880*x^3*y^2 -1591810452523410717496320*x^2*y^3 +6369821729789791023498240*x*y^4 -10195844311734842163382272*y^5 +698298992132204872800*x^4*z -22354621868063712556800*x^3*y*z +268364155878035524972800*x^2*y^2*z -1431855431362225199692800*x*y^3*z +2864871199378714763404800*y^4*z +628129563425241012000*x^3*z^2 -15081217751347682904000*x^2*y*z^2 +120698627643686966256000*x*y^2*z^2 -321993421558361015328000*y^3*z^2 +282505597812810990000*x^2*z^3 -4521921043759029720000*x*y*z^3 +18095013058413362040000*y^2*z^3 +63529419281243962500*x*z^4 -508441284296244225000*y*z^4 +5714558784617568750*z^5 +11357313312345614116464*x^4 -363581284514506596480384*x^3*y +4364743232413209659854464*x^2*y^2 -23288062757057189951768064*x*y^3 +46594997526073097237370624*y^4 +20432119573263584553120*x^3*z -490569561358714329935040*x^2*y*z +3926146667729872546546560*x*y^2*z -10473966637088120278145280*y^3*z +13784229776938358298600*x^2*z^2 -220637039346909605440800*x*y*z^2 +882905753885769555645600*y^2*z^2 +4133034860184110673000*x*z^3 -33077676077129884122000*y*z^3 +464715224352143724375*z^4 +221127342457388966072928*x^3 -5309206565905608012000576*x^2*y +42490862273001446293774464*x*y^2 -113354877311767229496844032*y^3 +298360636297854642229680*x^2*z -4775704447451559881135040*x*y*z +19110557991503891971673280*y^2*z +134189751333313878798600*x*z^2 -1073952985549277039720400*y*z^2 +20117588327927931586500*z^3 +2417220978290808419182908*x^2 -38691206453160492622123824*x*y +154827534413052279228563568*y^2 +2174323570423660733624820*x*z -17401636613860675368945480*y*z +488958501452646558817575*z^2 +14066144783900382791532852*x -112574753554133047365622728*y +6326345506535423378687790*z +34041853782739205900403193)
[y^3 -3,  z^2 -10*z +25,  x -1]
[x^2 -2,  y^3 -6*y^2 +12*y -8,  z -3]
true
true
ideal(
  x*z -3*x -z +3,
  x^2 +(-1/4)*z^2 +(5/2)*z -29/4,
  z^3 -13*z^2 +55*z -75,
  y^2*z^2 +(-2/3)*y^3 -10*y^2*z -2*y*z^2 +25*y^2 +20*y*z +(5/6)*z^2 -50*y +(-25/3)*z +137/6,
  y^3*z -3*y^3 -3*z +9,
  y^4 -4*y^3 +(19/4)*y*z^2 +(-95/2)*y*z -7*z^2 +(463/4)*y +70*z -163,
  x*y^3 -6*x*y^2 -y^3 +12*x*y +6*y^2 -8*x -12*y +8
)
[x^2 -2,  x -1]
x^3 -x^2 -2*x +2
3
12
[y^3 -3,  y -2]
y^4 -2*y^3 -3*y +6
4
8
[z -3,  z -5]
z^2 -8*z +15
2
5
ideal(
  x*z -3*x -z +3,
  x^2 +(-1/4)*z^2 +(5/2)*z -29/4,
  z^3 -13*z^2 +55*z -75,
  y^2*z^2 +(-2/3)*y^3 -10*y^2*z -2*y*z^2 +25*y^2 +20*y*z +(5/6)*z^2 -50*y +(-25/3)*z +137/6,
  y^3*z -3*y^3 -3*z +9,
  y^4 -4*y^3 +(19/4)*y*z^2 +(-95/2)*y*z -7*z^2 +(463/4)*y +70*z -163,
  x*y^3 -6*x*y^2 -y^3 +12*x*y +6*y^2 -8*x -12*y +8
)
28153056843*x^5 +267797369970*x^4*y +1018936334520*x^3*y^2 +1938464246160*x^2*y^3 +1843905014640*x*y^4 +701583371424*y^5 +521861541480*x^4*z +3971239047360*x^3*y*z +11332560208320*x^2*y^2*z +14373003191040*x*y^3*z +6835940542080*y^4*z +3869412405120*x^3*z^2 +22083963482880*x^2*y*z^2 +42013393943040*x*y^2*z^2 +26642640061440*y^3*z^2 +14345138672640*x^2*z^3 +54581503242240*x*y*z^3 +51918990888960*y^2*z^3 +26590988759040*x*z^4 +50587734712320*y*z^4 +19716245323776*z^5 -2490515540721*x^4 -18952215822072*x^3*y -54083152467864*x^2*y^2 -68593266544608*x*y^3 -32623626771216*y^4 -36932523140448*x^3*z -210785619874752*x^2*y*z -401006789030016*x*y^2*z -254296988165376*y^3*z -205380860390784*x^2*z^2 -781449127340544*x*y*z^2 -743329657714176*y^2*z^2 -507607980152832*x*z^3 -965693230534656*y*z^3 -470465932824576*z^4 +87712548496371*x^3 +500603325564654*x^2*y +952367302293732*x*y^2 +603940240478952*y^3 +975534685715736*x^2*z +3711790511503776*x*y*z +3530727559723104*y^2*z +3616616395824192*x*z^2 +6880392167665536*y*z^2 +4469314570449408*z^3 -1537225804497825*x^2 -5848956719552700*x*y -5563641757623300*y^2 -11397966940666800*x*z -21683937106634400*y*z -21127938719284800*z^2 +13402584284506182*x +25497599370523956*y +49687629542559504*z -46482784771437882
5
5
